### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [77]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [78]:
# Extract unique players SN.
players = purchase_data['SN'].unique()

# Create a dictionary to store players' count
myDict=[{"Total Players": players.size}]

total_players = pd.DataFrame(myDict)
total_players 




Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [79]:
unique_items = purchase_data['Item Name'].unique()
items_count = unique_items.size

avarage_price = round(purchase_data['Price'].mean(),2)

purchase_count = purchase_data.index.stop

total_revenue = round(purchase_data['Price'].sum(),2)

purchase_analysis = pd.DataFrame([
    {'Number of Unique Items':items_count,
     'Average Price':avarage_price,
        'Number of Purchases':purchase_count,
     'Total Revenue':total_revenue
    }
])


# Format cells
purchase_analysis['Average Price'] = purchase_analysis['Average Price'].map("${:,.2f}".format)
purchase_analysis['Total Revenue'] = purchase_analysis['Total Revenue'].map("${:,.2f}".format)


purchase_analysis

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [80]:

# Create a base table with duplicate SN's and respective genders
uniplayers_df = purchase_data.loc[:,['SN', 'Gender']]

# Eliminate duplicate rows
uniplayers_df.drop_duplicates(inplace=True)

# Group SNs by Gender and aggregate count total Sn's under each gender
genders_df = uniplayers_df.groupby('Gender').count()

# Rename 'SN' column header to 'Total Count'
genders_df.rename(columns = {'SN':'Total Count'},inplace= True)

# Add 'Percentage of Players' column and calculate precentages
genders_df['Percentage of Players'] = round((genders_df['Total Count']/ genders_df['Total Count'].sum())*100,2)

# Format the 'Percentage of Players' column
genders_df['Percentage of Players'] = genders_df['Percentage of Players'].map("{:,.2f}%".format)

#Show ouput dataframe
genders_df


Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [81]:

PAG_df = purchase_data.loc[:,['Gender','Price']].groupby('Gender').agg(['count','mean', 'sum'])


PAG_df= pd.merge(genders_df,PAG_df,on = 'Gender')

PAG_df.columns

PAG_df['Avg Total Purchase per Person'] = round(PAG_df[('Price', 'sum')] / PAG_df['Total Count'],2)

PAG_df.drop(axis=1, inplace= True, columns = ['Total Count','Percentage of Players'])

PAG_df.rename(inplace = True,columns = {('Price', 'count'): 'Purchase Count', 
                         ('Price', 'mean'): 'Average Purchase Price',
                         ('Price', 'sum'):'Total Purchase Value'
                        })

PAG_df['Average Purchase Price'] = PAG_df['Average Purchase Price'].map("${:,.2f}".format)
PAG_df['Total Purchase Value'] = PAG_df['Total Purchase Value'].map("${:,.2f}".format)
PAG_df['Avg Total Purchase per Person'] = PAG_df['Avg Total Purchase per Person'].map("${:,.2f}".format)


PAG_df

C:\Users\spaki\anaconda3\lib\site-packages\pandas\core\reshape\merge.py:618: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)
  warnings.warn(msg, UserWarning)


Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [82]:
# Create Base table.
AD_df = purchase_data.loc[:,['SN','Age']]

# Drop duplicate rows.
AD_df.drop_duplicates(inplace=True)

# Drop redundant column - SN.
AD_df.drop(columns=['SN'], inplace=True)

# Create age group bins.
bins = [0, 9, 14, 19, 24, 29, 34,39, 100]

# Create age group labels.
mylabels = ['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']

# Add column with respective age groups.
AD_df['Age Group'] = pd.cut(AD_df['Age'], bins=bins, labels = mylabels,include_lowest=True)

# Group by age group and count each age group.
AD_df=AD_df.groupby('Age Group').count()

# Add column with respective age groups percentages.
AD_df['Percentage of Players'] = round(AD_df['Age']/AD_df['Age'].sum()*100,2)

# Polish and format table
AD_df.rename(inplace= True, columns = {'Age': 'Total Count'})

AD_df['Percentage of Players'] = AD_df['Percentage of Players'].map("{:,.2f}%".format)

AD_df

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [83]:

PAA_df = purchase_data.loc[:,['Age','Price']]

PAA_df['Age Group'] = pd.cut(PAA_df['Age'], bins=bins, labels = mylabels,include_lowest=True)

PAA_df.drop(columns= 'Age',inplace=True)

PAA_df = PAA_df.groupby('Age Group').agg(['count','mean', 'sum'])

PAA_df.columns = PAA_df.columns.droplevel(0)

PAA_df= pd.merge(AD_df,PAA_df,on = 'Age Group')


PAA_df['Avg Total Purchase per Person'] = round(PAA_df[('sum')] / PAA_df['Total Count'],2)

PAA_df.drop(axis=1, inplace= True, columns = ['Total Count','Percentage of Players'])



PAA_df.rename(inplace = True,columns = {'count': 'Purchase Count', 
                         'mean': 'Average Purchase Price',
                         'sum':'Total Purchase Value'
                         })

#PAA_df['Avg Total Purchase per Person'] = round(PAA_df[('Total Purchase Value')] / PAA_df['Purchase Count'],2)

PAA_df['Average Purchase Price'] = PAA_df['Average Purchase Price'].map("${:,.2f}".format)
PAA_df['Total Purchase Value'] = PAA_df['Total Purchase Value'].map("${:,.2f}".format)
PAA_df['Avg Total Purchase per Person'] = PAA_df['Avg Total Purchase per Person'].map("${:,.2f}".format)


PAA_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Avg Total Purchase per Person  
Age Group                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [84]:
TS_df = purchase_data.loc[:,['SN','Price']].groupby('SN').agg(['count','mean', 'sum'])


TS_df.columns=TS_df.columns.droplevel(0)

TS_df.sort_values('sum',ascending = False,inplace= True)

TS_df.rename(inplace = True, columns = {'count':'Purchase Count',
                                        'mean':'Average Purchase Price',
                                       'sum':'Total Purchase Value'})

TS_df['Average Purchase Price'] = TS_df['Average Purchase Price'].map("${:,.2f}".format)
TS_df['Total Purchase Value'] = TS_df['Total Purchase Value'].map("${:,.2f}".format)

TS_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [88]:
input_df = purchase_data.loc[:,['Item ID','Item Name','Price']]


MPI_df = input_df.groupby([ 'Item ID','Item Name']).agg(['count','min','sum'])

MPI_df.rename(inplace= True, columns = {'count':'Purchase Count' ,'min':'Item Price','sum':'Total Purchase Value'}) 

MPI_df.columns=MPI_df.columns.droplevel(0)

MPS_df = MPI_df.copy()

MPI_df['Item Price'] = MPI_df['Item Price'].map("${:,.2f}".format)
MPI_df['Total Purchase Value'] = MPI_df['Total Purchase Value'].map("${:,.2f}".format)


MPI_df.sort_values('Purchase Count',ascending = False,inplace= True)

MPI_df.head()


Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.19   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.19   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [86]:
MPS_df.sort_values('Total Purchase Value',ascending = False,inplace= True)

MPS_df['Item Price'] = MPS_df['Item Price'].map("${:,.2f}".format)
MPS_df['Total Purchase Value'] = MPS_df['Total Purchase Value'].map("${:,.2f}".format)

MPS_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.19   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Observations

* 1. Analysis shows that spending pattern of individuals are not significantly sensitive to neither their age group, sex, or age. 


* 2. it is important to remember that nearly 85% of the players are males. Also, more than 77% of individuals are between in between 15 and 25 years old. The trend quickly winds down in the individuals within their 30’s and older. Therefore, the game may find better audience within male collage students.


* 3. The 2 most profitable and most popular items are the same. Also, it is observed that Price does not significantly impact the popularity of an item in a negative way. Therefore, more expensive items may easily result in better profitability.
